fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -3,204 +3,541 @@
+
+        if dist_record is not None and dist_record.rank >= 0:
+            for e in dist_record.batch_events[chunk]:
+                if e is not None and is_cuda(self.stream):
+                    self.stream.wait_event(e)
+
+        # Determine whether checkpointing or not.
+        checkpoint = chunk < checkpoint_stop
+        if checkpoint:
+
+            def function(input: TensorOrTensors, chunk_id: int = chunk) -> TensorOrTensors:
+                with record_function("chunk%d-rank%d" % (chunk_id, dist_record.rank)):
+                    result = self.module(*input)
+                    if self.num_output is None:
+                        result = (result,)
+                    return tuple(result)
+
+            chk = Checkpointing(function, batch)
+            task = Task(self.stream, compute=chk.checkpoint, finalize=chk.recompute)
+            del function, chk
+
+        else:
+
+            def compute(
+                batch: Batch = batch,
+                chunk_id: int = chunk,
+                rank: int = dist_record.rank if dist_record is not None else -1,
+            ) -> Batch:
+                with record_function("chunk%d-rank%d" % (chunk_id, dist_record.rank)):
+                    result = self.module(*batch.tensors)
+                    if self.num_output is None:
+                        result = (result,)
+                return Batch(result, chunk_id)
+
+            task = Task(self.stream, compute=compute, finalize=None)
+            del compute
+
+        self.in_queue.put(task)
+
+        ok, payload = self.out_queue.get()
+
+        # Hold the first exception.
+        if exc_info is not None:
+            pass
+        elif not ok:
+            exc_info = cast(ExcInfo, payload)
+        else:
+            task, batch = cast(Tuple[Task, Batch], payload)
+
+            with use_device(self.device):
+                task.finalize(batch)
+
+            dist_record.batches[chunk] = batch
+
+        if exc_info is not None:
+            raise exc_info[0].with_traceback(exc_info[1], exc_info[2])
+
+    def forward_results(self, chunk: int, dist_record: DistributedPipelineRecord) -> None:
+        with use_stream(self.stream):
+            for user, input_idx, output_idx in dist_record.users:
+                v = dist_record.get_batch(chunk).value[output_idx]
+                if BOUNCE_TENSORS:
+                    v = v.cpu()
+                dist_record.forwarded_phony[chunk][output_idx].append(user.remote().feed(chunk, input_idx, v))
+
+    def run_pipeline(self, dist_record_ref: rpc.RRef) -> Optional[Tensor]:
+        dist_record = dist_record_ref.local_value()
+        self.run(dist_record)
+
+        if not dist_record.users:
+            result = microbatch.gather(dist_record.batches)
+            assert len(result) == 1
+            result = result[0]
+            s0 = current_stream(result.device)
+            if is_cuda(s0):
+                as_cuda(s0).synchronize()
```

Comment:
Why is this synchronize necessary? It will not be good for performance.

Reply:
Adding a TODO to revisit that. I believe this was needed, but needs investigation why.